fix(docs): rewrote readme (#1172)
blackflux/lambda-serverless-api: README.md

[![Semantic-Release](https://github.com/blackflux/js-gardener/blob/master/assets/icons/semver.svg)](https://github.com/semantic-release/semantic-release)
[![Gardener](https://github.com/blackflux/js-gardener/blob/master/assets/badge.svg)](https://github.com/blackflux/js-gardener)

Middleware for AWS Lambda and Api Gateway.

Provides support for:

> - Fully customizable parameters types

Reviewer: ```suggestion
- Fully customizable parameter types
```

> - Support for `header`, `json`, `query` and `context` parameter positions 

Reviewer: ```suggestion
- Support for `header`, `json`, `query` and `context`, parameter positions 
```

> - Abstraction for `text`, `json` and `binary` response types

Reviewer: ```suggestion
- Abstraction for `text`, `json`, and `binary` response types
```

- Full support for custom authentication
- Full support for `cors`
- Automatic generation of [Swagger](https://swagger.io/) documentation
- Rate limiting using [lambda-rate-limiter](https://github.com/blackflux/lambda-rate-limiter)
- Access and error logging using [lambda-monitor](https://github.com/blackflux/lambda-monitor)

## Install

    $ npm install --save lambda-serverless-api

## Getting Started

> Define api and handlers in `handler.js` as following

Reviewer: ```suggestion
Define api and handlers in `handler.js` as following:
```

<!-- eslint-disable import/no-unresolved -->
```js
const api = require('lambda-serverless-api').Api({/* options */});

api.wrap('POST register', [
  api.Str('name', 'json', { required: false }),
  api.Email('email', 'json'),
  api.Str('password', 'json')
], /* options, */ ({ name, email, password }) => {
  // handle registration logic here ...
  if (new Date().getHours() === 4) {
    throw api.ApiError('I am a teapot', 418);
  }
  return api.JsonResponse({ message: 'Success!' });
});

module.exports = api;

```

then hook the router endpoint into the [Serverless Framework](https://serverless.com/) configuration as

```yaml
functions:
  router:
    handler: handler.router
    events:
      - schedule: rate(10 minutes)
      - http:
          path: /{proxy+}
          method: ANY
```

## Api Options

> The api is plugin based and all options are tied to plugins. The following plugins are customizable

Reviewer: ```suggestion
The api is plugin based and all options are tied to plugins. The following plugins are customizable:
```

- _cors_: Used to handle CORS Option requests as well as injecting CORS headers into responses

> - _logger_: Log responses to console / cloudwatch, which can then be picked up by `lambda-monitor`

Reviewer: ```suggestion
- _logger_: Log responses to console (cloudwatch), which can then be picked up by `lambda-monitor`
```

> - _preValidation_: Hook to allow pre validation (e.g. authentication)

Reviewer: ```suggestion
- _preValidation_: Hook to allow pre-validation (e.g. authentication)
```

- _rateLimit_: Used for rate limiting 
- _router_: Used to modify the router, e.g. set a custom route prefix

Please see implementation for details.

## Endpoint Definition: wrap()

Takes the following positional arguments:

> - _route_: A string consisting of the method and the uri

Reviewer: ```suggestion
- _route_ `string`: The method and the uri of the format `${method} ${uri}`
```

> - _params_: An array containing the parameter definitions

Reviewer: ```suggestion
- _params_ `Array<Parameter>`: [Parameter definitions](#api-parameters)
```

> - _options_: Endpoint options (optional)

Reviewer: ```suggestion
- _options_ `object`: [Endpoint options (optional)](#parameter-options)
```

> - _handler_: The handler taking as first argument the parsed parameters

Reviewer: ```suggestion
- _handler_ `function`: The handler using parsed parameters as the first parameter
```

});
```

### Parameter Names

Parameter names are converted to camel case when passed into the handler.

E.g. the header `X-Custom-Header` would be passed as `xCustomHeader`.

## Api Response

There are two types of api responses: Success and error responses. Error responses need to be thrown, while success responses
need to be returned from the endpoint handler. The following responses are available:

- _Success_: `ApiResponse` (e.g. for plain text), `JsonResponse` (for json), `BinaryResponse` (e.g. for images)
- _Failure_: `ApiError`

Success responses take the positional parameters `data` (required), `statusCode` (optional) and `headers` (optional).

Error responses take the positional parameters `message` (required), `statusCode` (optional), `messageId` (optional) and `context` (optional). They are always of Json format. It is recommended that an easily identifiable messageId is set on all error responses.

Please see implementation for details.

## Rate Limiting

> Done globally through the API configuration. However can be overwritten on an per-endpoint basis

Reviewer: ```suggestion
Done globally through the API configuration. However can be overwritten on a per-endpoint basis
```
